src/nvim/normal.c

```diff
@@ -3611,6 +3625,107 @@ void scroll_redraw(int up, long count)
   redraw_later(VALID);
 }
 
+static linenr_T longest_lnum;
+/*
+ * Do a horizontal scroll.  Return TRUE if the cursor moved, FALSE otherwise.
+ */
+int gui_do_horiz_scroll(int leftcol, bool compute_longest_lnum)
+{
+    /* no wrapping, no scrolling */
+    if (curwin->w_p_wrap)
+      return false;
+
+    if (curwin->w_leftcol == (colnr_T)leftcol)
+      return false;
+
+    curwin->w_leftcol = (colnr_T)leftcol;
+
+    /* When the line of the cursor is too short, move the cursor to the
+     * longest visible line. */
+    if (!virtual_active()
+      && (colnr_T)leftcol > scroll_line_len(curwin->w_cursor.lnum))
+    {
+      if (compute_longest_lnum)
+      {
+          curwin->w_cursor.lnum = gui_find_longest_lnum();
+          curwin->w_cursor.col = 0;
+      } else if (longest_lnum >= curwin->w_topline
+        && longest_lnum < curwin->w_botline) {
+        /* Do a sanity check on "longest_lnum", just in case. */
+        curwin->w_cursor.lnum = longest_lnum;
+        curwin->w_cursor.col = 0;
+      }
+    }
+
+    return leftcol_changed();
+}
+
+/*
+ * Find longest visible line number.  If this is not possible (or not desired,
+ * by setting 'h' in "guioptions") then the current line number is returned.
+ */
+static linenr_T gui_find_longest_lnum()
+{
+    linenr_T ret = 0;
+
+    /* Calculate maximum for horizontal scrollbar.  Check for reasonable
+     * line numbers, topline and botline can be invalid when displaying is
+     * postponed. */
+    if (curwin->w_topline <= curwin->w_cursor.lnum
+      && curwin->w_botline > curwin->w_cursor.lnum
+      && curwin->w_botline <= curbuf->b_ml.ml_line_count + 1)
+    {
+      linenr_T    lnum;
+      colnr_T     n;
+      long      max = 0;
+
+      /* Use maximum of all visible lines.  Remember the lnum of the
+       * longest line, closest to the cursor line.  Used when scrolling
+       * below. */
+      for (lnum = curwin->w_topline; lnum < curwin->w_botline; ++lnum)
+      {
+          n = scroll_line_len(lnum);
+          if (n > (colnr_T)max)
+          {
+            max = n;
+            ret = lnum;
+          } else if (n == (colnr_T)max
+            && abs((int)(lnum - curwin->w_cursor.lnum))
+               < abs((int)(ret - curwin->w_cursor.lnum))) {
+              ret = lnum;
+          }
+      }
+    } else {
+      /* Use cursor line only. */
+      ret = curwin->w_cursor.lnum;
+    }
+
+    return ret;
+}
+
+/*
+ * Return length of line "lnum" for horizontal scrolling.
+ */
+static colnr_T scroll_line_len(linenr_T lnum)
+{
+    char_u  *p;
+    colnr_T col;
+    int   w;
+
+    p = ml_get(lnum);
+    col = 0;
+    if (*p != NUL)
+      for (;;)
+      {
```

Comment:
That is the way that it is written in the VIM source. If necessary, I can modify it to be less ambiguous but I tried to leave the code as close as possible to the original.